src/abstractions/store/InMemoryBackingStore.cs

```diff
@@ -128,6 +142,14 @@ public void Set<T>(string key, T? value)
         /// <returns>A collection of strings containing keys changed to null </returns>
         public IEnumerable<string> EnumerateKeysForValuesChangedToNull()
         {
+            if(ReturnOnlyChangedValues) // refresh the state of collection properties if they've changed in size.
+            {
+                foreach(var item in store)
```

Comment:
we'd be enumerating the store twice, what do you think of moving the condition in the loop below instead?

Reply:
It turns out there's a reason why `Enumerate()` uses 2 consecutive iterations.

Since `EnsureCollectionPropertyIsConsistent()` does not simply flip the dirty flag in-place but instead calls `.Set()`, which will replace the object in the store entirely, performing both operations in a single loop doesn't actually work as the store will be properly updated but the current iteration variable will stay the same. Therefore, if we use a single iteration, an item that should be detected as dirty will be properly updated in the store but will not be considered in the current enumeration as dirty.

A quick check gives me the feeling that the call to `.Set()` is done to properly support the event subscription, but I think it might be possible to change its implementation so it reuses/updates an existing value in the store rather than recreating a new one from scratch every time (this would probably improve memory allocation overall, too). Still, I think we're starting to stray from the goal of this PR here.

Do you want me to push the research further on how `.Set()` could be improved to support a single iteration here (and potentially save memory allocations when ensuring the collection consistency) or can we consider this PR as good enough?

Reply:
> I don't mind moving the check in the same loop, but this pattern was based on the current implementation of `Enumerate()` right above it, which also enumerates the store twice.
> 
> Should I move both?

yes please, and enumerate could even yield return instead of building an intermediary list.

> Do you want me to push the research further on how .Set() could be improved to support a single iteration here (and potentially save memory allocations when ensuring the collection consistency) or can we consider this PR as good enough?

yes please, but timebox it. I don't want you to burn too much time on this.

Reply:
> yes please, but timebox it. I don't want you to burn too much time on this.
I've outran my timebox on this. For it to work, it basically requires a complete rewrite.

First the store itself needs to change, as it currently runs on immutable Tuples which are forcing to reassign a new object every time something changes, preventing a simple enumeration since the source is constantly modified. Switching to mutable objects with the intent to update them whenever possible means the `.Set()` function also needs to be rewritten, along with every check for Tuple/collections. At this point most of the class needs to be modified. This is a rather involved change and I'm not quite sure the level of effort/risk required is worth the gain - particularly in the scope of this PR.

> yes please, and enumerate could even yield return instead of building an intermediary list.
This could be a breaking change. The function is part of the interface, so anyone could currently be calling `Enumerate()` and changing the content of the store between yields, leading to undefined behavior. With the current implementation this works because of the intermediary list. I don't mind changing this, as long as we're aware of the consequences. On the positive side, it would allocate much less memory when iterating recursively within `EnsureCollectionPropertyIsConsistent()`.

Reply:
thanks for the detailed information and for spending the time investigating this. Resolving for now.